design_drafts/async_controller_impl.md

```diff
@@ -0,0 +1,40 @@
+
+#### Implementation
+
+The controllers running asynchronously are entirely independent of each other, which means that no data sharing occurs between them. Therefore synchronization happens between the Controller Manager's thread and an async controller - so in the following drawings only these two threads will be visualized.
+
+The critical sections are:
+
+* The state interface writes of the Controller Manager's read function
+* The command interface writes of the async controller's update function
+
+The former has to occur and be seen by the async thread before the update, while the latter, ideally, has to finish and be seen before the controller manager's write.
+
+![async_controller](https://user-images.githubusercontent.com/25421074/212979058-5dad88c1-d940-4bc7-ac6f-103711de7210.png)
+
+As shown in the above image, there are four different cases to account for, and these are: 
+
+1. The state interfaces aren't written yet by the cm read. In this case, we can be sure that the release flag isn't set, thus the acquire load on the other side will return false - so we'll skip the update and the asynchronous controller will sleep until the next iteration. This works as intended.
+
+2. The state interfaces are written in the cm read function, and the release flag is set. It can happen that the state interfaces are visible on the controller's side, but the flag isn't - in this case, we'll do the same as the first scenario. On the other hand, if the flag is set, then the thread's acquire load will return true if it isn't currently sleeping, and the respective orderings guarantee that the async update will be able to work with the latest state interface values. This is fine too.
+
+3. After async update, a release store is made to signal that the command interfaces are written. Again, the release ordering guarantees that if the store is visible on the other thread, then so are the command interfaces. If the acquire load of this variable is true right before the cm write function, then we're again working with the latest values, which is the intended outcome.
+
+4. If the flag isn't visible, and the load return false, it means that the asynchronous update isn't finished before the next write. This is not ideal, since it means that two threads might access the same variable in the same time, leading to a potential data race. Currently, if this happens, only a warning is given to the user, and we continue with the write function as if nothing happened. In practice, probably nothing dangerous will happen, since the aligned loads and stores of plain pointers (similarly to ints and bools) are atomic on most CPUs, and this can only occur if the async update is slower than all synchronous updates combined, but even then, this is an issue and should be handled differently.
```

Comment:
Which "load" are you talking about?

Reply:
"data race" --> "race condition"

Reply:
Not sure I understand this issue, sorry.

Reply:
Yes, this was not very clear. Added which load I was actually referring to